### ISSUE TITLE
EpochManager::get_epoch_validator_info has bug prone semantics

### ISSUE THREAD

**Issue by matklad** [near/nearcore#6572]
EpochManager is essentially a bunch of caches over a database. It also is mutable, and reflects mutations in the cache, even before they hit the database. 

`get_epoch_validator_info` breaks this pattern – it doesn't have a cache, so it'll return stale results after mutation.

**Comment by matklad** [comment]
https://github.com/near/nearcore/pull/6549#discussion_r847173102

**Comment by matklad** [comment]
Ok, so it's actually worse than that -- `get_epoch_validator_info` is the one that is "correct" (with respect to some definition of correct), while all the others are wrong. 

The overall flow of data here is that `EpochManager` doesn't actually write to the database directly (outside of constructor). Instead, it returns a store update (essentially, a diff to be applied to the database). That `StoreUpdate` gets routed, via `RuntimeAdapter::add_validator_proposals`, to `ChainUpdate::process_block`, where the update actually gets committed or aborted. That is, epoch manager caches optimistically apply TX without checking whether tx actually commits.

**Comment by bowenwang1996** [comment]
> epoch manager caches optimistically apply TX without checking whether tx actually commits

Yes this is a known suboptimality. Related issue #4243

**Comment by mzhangmzz** [comment]
> Ok, so it's actually worse than that -- `get_epoch_validator_info` is the one that is "correct" (with respect to some definition of correct), while all the others are wrong.
> 
> The overall flow of data here is that `EpochManager` doesn't actually write to the database directly (outside of constructor). Instead, it returns a store update (essentially, a diff to be applied to the database). That `StoreUpdate` gets routed, via `RuntimeAdapter::add_validator_proposals`, to `ChainUpdate::process_block`, where the update actually gets committed or aborted. That is, epoch manager caches optimistically apply TX without checking whether tx actually commits.

Good point! What do you think is the best way to fix this @matklad? It seems like to fix this completely we need a big change in EpochManager. Since we are thinking about refactoring store in general, should we delay this fix until we figure out the plan there?

**Comment by matklad** [comment]
I definitely don't see any easy fixes here :-) FWIW, at the moment I am not even sure what the hard fix would be
